xapian-applications/omega/clickmodel/docs/clickmodel.rst

```diff
@@ -0,0 +1,66 @@
+.. Copyright (C) 2017 Vivek Pal
+
+===============================================
+Simplified Dynamic Bayesian Network Click Model
+===============================================
+
+.. contents:: Table of Contents
+
+Introduction
+============
+
+Clicks in web search reflect user interests in retrieved documents and by the
+use of various methods, clicks can be used to improve search results. Click
+models are one such methods. These are probabilistic graphical models that
+make user clicks easier to understand, define and quantify.
+
+There are various kinds of click models among which the dynamic bayesian
+network (DBN) model is proved to be able to closely mimic the user’s search
+behavior. The underlying idea of DBN model is that if a user is satisfied
+with a document in the search engine results page (SERP), he/she stops
+otherwise, continues to examine more documents with fixed probability.
+
+The Simplified DBN model (SDBN), according to the paper_: `Chapelle, Olivier and
+Zhang, Ya. A dynamic bayesian network click model for web search ranking.
+Proceedings of WWW, pages 1-10, 2009.`__
+is a simplified variation of DBN model wherein the inference algorithm is
+relatively simpler. It contains the set of attractiveness and satisfactoriness
+parameters, which both depend on a query and a document.
+
+.. _paper: http://dl.acm.org/citation.cfm?id=1526711
+__ paper_
+
+Training the SDBN model
+=======================
+
+For training the model, we need the list of search sessions obtained from
+clickstream log data. Xapian Omega has the clickstream logging functionality
+which generates click log data in the required format in ``final.log`` file.
+
+A few example entries in ``final.log``:
+
+	QueryID,Query,Hits,Offset,Clicks
+	821f03288846297c2cf43c34766a38f7,book,"45,54",0,"45:0,54:2"
+	098f6bcd4621d373cade4e832627b4f6,test,"35,47",0,"35:1,47:0"
+
+``SimplifiedDBN`` class provides train function that takes a search session
+as its input:
+
+	``void SimplifiedDBN::train(vector<vector<string>> &sessions)``
+
+To generate search sessions from the log data file, ``SimplifiedDBN`` class
+also provides a separate function which takes the path to the log data file
+as its input:
+
+	``vector<vector<string>> SimplifiedDBN::buildSessions(string logfile)``
+
+Predicting the Document Relevance
+=================================
+
+``SimplifiedDBN`` class provides a function to get the predicted relevance of
+all documents in a search session based on the trained SDBN model:
+
+	vector<int> get_predicted_relevances(vector<vector<string>> &sessions);
+
+It takes a search session as its input and returns a list of predicted relevances
+of each document in that search session.
```

Comment:
"It takes a search session as its input and returns a list of predicted relevances of each document in that search session" doesn't really tell me anything very useful.

The reader can see from just looking at the method prototype that you pass it "sessions"; also the method name is `get_predicted_relevances()` and it returns `vector<int>`, so it's expected that the returned value is a list of predicted relevances.

To be useful, you need to tell the reader things they can't trivially work out for themselves from the names of the function and its parameters.

For example, how does one create a suitable `vector<vector<string>>` value to pass as the argument?  I'm guessing from the `build_sessions()` method, but I shouldn't need to guess.

And what are the values in the returned `vector<int>` and what would one do with them?  Despite reading through the code, I've no idea about that.  Without knowing that, I can't usefully write code to use this method.

I also think it would be better to put this documentation as extended documentation comments for the corresponding methods in the header file.  Currently the header file just has a brief description for each method, then the user needs to somehow know to come and look here to find more detailed information.

Also, splitting the information up like this makes it more likely that only one part gets updated if a method signature is changed.  Documentation comments in the headers work well because people modifying the method signature are more likely to see they need updating.  As a case in point, the documentation above still says `vector<vector<string>> SimplifiedDBN::buildSessions(string logfile)`, but the header currently has `vector<vector<string>> build_sessions(const string logfile);`.